### PR TITLE
Integrate Wasmer into Substrate sandbox environment

### PR DESCRIPTION
**Pull request by 0x7CFE**
## TL;DR

Currently Substrate can execute Wasm in two ways: either by interpreting it using Wasmi or by compiling Wasm bytecode into native instructions using Wasmtime. The latter is faster to execute, but is only supported for trusted code, i.e. runtime. For security reasons smart contracts are still interpreted.

In order to execute smart contracts safely we need to have an environment that is suitable to execute untrusted code and resistant to JIT bomb attacks. Such an environment is provided by the [Wasmer Singlepass](https://crates.io/crates/wasmer-compiler-singlepass) compiler that is advertised as being designed specifically with blockchain in mind:

> Singlepass is designed to emit compiled code at linear time, as such is **not prone to JIT bombs** and also offers great compilation performance orders of magnitude faster than wasmer-compiler-cranelift and wasmer-compiler-llvm, however with a bit slower runtime speed.
> 
> The fact that singlepass is not prone to JIT bombs and offers a very predictable compilation speed makes it **ideal for blockchains** and other systems where fast and consistent compilation times are very critical.

This PR integrates Wasmer Singlepass into Substrate sandbox environment.

## API conundrum

Wasmer has rather simple API so its integration was straightforward. However, Substrate sandbox API was designed and implemented with Wasmi in mind, which is essentially single-threaded. This is incompatible with that of Wasmer and really annoys Rust borrowck.

Namely, Wasmer forces host function closures to be `Fn + Send + Sync`. Yet,
- `supervisor_externals` is passed as `&mut`
- `dispatch_thunk` is neither `Send` nor `Sync`

Both of those entities are used by host function closures to dispatch the call.

## What was done

The solution to both issues is scoped thread local storage. Since current execution model is single-threaded anyway, introducing TLS does not violate invariants and calms down the borrow checker. This was done by carefully wrapping objects in a scoped TLS and making state objects clonable by wrapping inner data with `Rc`.

Also, please note that there are a couple of `unsafe` one-liners to access construct a slice that views into wasmer memory. The context is very limited and single-threaded, so I suppose this is safe.

## Memory API redesign

To eliminate potential aliasing issues when accessing memory, the sandbox memory access API was redesigned. It is now use external buffers. Also the `MemoryTransfer` trait was introduced to abstract over backends.

## How it was tested

I successfully tested it on the following ink! examples:
- `flipper`
- `incrementer`
- `erc20`
- `erc721`
- `dns`
- `delegator`

Considering sufficiently complex logic of example contracts (such as hash table manipulation and cross module operation) I cautiously state that the code is working as expected and is ready for extensive testing in a real world scenarios.

When testing yourself don't forget to pass `--features wastime,wasmer-sandbox` to `cargo build` and `--execution Wasm --wasm-execution Compiled` when running a node.

## Further work

Current implementation hard-codes sandbox execution mode to match that of runtime and is gated by `wasmer-sandbox` feature flag. So, if control flow passing from runtime being interpreted by Wasmi, then sandboxed module will be interpreted by Wasmi as well. And vice versa, if runtime is executed by Wasmtime, then sandbox will be executed by Wasmer. If `wasmer-sandbox` feature is disabled, sandbox will always be executed via wasmi, as it was before.

In some cases there might be an intention to force execution mode of a sandboxed code, say, for security or performance reasons. In that case command line parameter is needed. I tried adding one, but unfortunately this ended up being a non trivial task, since currently there is no way to pass configuration options to the sandbox environment which is instantiated on a per call basis. Of course, this is possible but would require substantial refactoring, which probably should be done in a separate PR.

Also we should refactor sandbox logic to eliminate copy-pasta in Value conversion. Yet, I'd really like to do that in a separate PR.

## Note to reviewers

This PR affects execution of untrusted code and contains `unsafe` one-liners. Please review very carefully. I may suggest starting from `common/src/sandbox.rs` and `common/src/util.rs`.

### CODE REVIEW

**Comment by 0x7CFE** [comment]
[clabot:check]

**Comment by athei** [comment]
Just by looking at the changed files: Is the wasmer support implemented inside `executor-wasmtime`?  Does this crate now contain two different wasm executors or does it replace wastime?

File: client/executor/wasmtime/src/host.rs [hunk 1]
```diff
@@ -315,20 +405,31 @@ impl<'a> Sandbox for HostContext<'a> {
...
-			};
+		let guest_env = match sandbox::GuestEnvironment::decode(
+			&*self.inner.sandbox_store.borrow(),
+			raw_env_def,
+		) {
+			Ok(guest_env) => guest_env,
+			Err(_) => return Ok(sandbox_primitives::ERR_MODULE as u32),
+		};
+
+		println!("Instantiating sandbox from host");
```
Comment (athei) [inline client/executor/wasmtime/src/host.rs:416]: Some leftover debugging code.

Comment (0x7CFE) [reply client/executor/wasmtime/src/host.rs:416]: Missed that, thanks!

**Comment by 0x7CFE** [comment]
> Just by looking at the changed files: Is the wasmer support implemented inside `executor-wasmtime`? Does this crate now contain two different wasm executors or does it replace wastime?

The branch name is a misnomer due to historic reasons. This branch originally was created to support Lightbeam as a singlepass backend for wasmtime. For that sandbox was altered to support wasmtime execution, then Lightbeam branch was added, then it was all removed in favor of Wasmer's singlepass compiler.  

Most of the implementation details are specific to sandbox execution, so the "meat" is in `sandbox.rs`. However, trusted wasm logic is still executed by wasmtime. Where runtime calls into sandbox this flow is delegated either to Wasmi or to Wasmer Singlepass depending on the `wasmer-sandbox` feature flag.

**Comment by athei** [comment]
Can you rename `executor-wasmtime` to no longer be a misnomer? Like `executor-wasm-compiled` or something? Or should they be rather different crates?

**Comment by 0x7CFE** [comment]
> Can you rename `executor-wasmtime` to no longer be a misnomer? Like `executor-wasm-compiled` or something? Or should they rather be different crates?

`executor-wasmtime` remains intact. It still contains wasmtime. The PR branch name was a misnomer, not the crate.

**Comment by athei** [comment]
> > Can you rename `executor-wasmtime` to no longer be a misnomer? Like `executor-wasm-compiled` or something? Or should they rather be different crates?
> 
> `executor-wasmtime` remains intact. It still contains wasmtime. The PR branch name was a misnomer, not the crate.

Yes my bad. Got that now. I was confused by the fact that the `wasmer-sandbox` feature is passed through to `executor-wasmtime`. Isn't that a breach of encapsulation when the sandbox is implemented in `executor-common`? Can't we get away with the specific executor not knowing how the sandbox code is executed? I know it might be hard but it would be much nicer if we could abstract over the API differences there.

File: client/executor/wasmi/src/lib.rs [hunk 2]
```diff
@@ -261,13 +390,47 @@ impl<'a> Sandbox for FunctionExecutor<'a> {
...
 			.map(|i| i.get_global_val(name))
 			.map_err(|e| e.to_string())
 	}
 }
 
+struct CapsHolder;
+
+scoped_tls::scoped_thread_local!(static EXECUTOR: FunctionExecutor);
+
+impl sandbox::SandboxCapabiliesHolder for CapsHolder {
```
Comment (apopiak) [inline client/executor/wasmi/src/lib.rs:407]:
```suggestion
impl sandbox::SandboxCapabilitiesHolder for CapsHolder {
```
typo

**Comment by 0x7CFE** [comment]
> Yes my bad. Got that now. I was confused by the fact that the `wasmer-sandbox` feature is passed through to `executor-wasmtime`. Isn't that a breach of encapsulation when the sandbox is implemented in `executor-common`? Can't we get away with the specific executor not knowing how the sandbox code is executed? I know it might be hard but it would be much nicer if we could abstract over the API differences there.

Currently sandbox dispatch mode (wasmi or wasmer) is defined at compile time using that feature flag. Theoretically nothing prevents one from selecting dispatch mode on a per call basis, except that current executor design is not very flexible and does not propagate settings into its internals. So there is no was to know what command line argument was passed from inside the `wasmtime/host.rs`.

Right now `wasmer-sandbox` feature flag that dictates what sandbox backend would be instantiated:

https://github.com/paritytech/substrate/blob/e3d09fc1795cf5dbcad55d6fc4d9588f8eb379a7/client/executor/wasmtime/src/host.rs#L68-L80

**Review by pepyakin — CHANGES_REQUESTED** [review]
I had a first approach at this PR and my first impression that this could use some refactoring. Right now it seems to me that the new logic is plopped onto existing code.

Here are some examples, not exhaustive:

- `instantiate` is giant and concentrates lots of different concerns.
- unsafe access to the guest memory is spreaded across the codebase. This could really use an additional abstraction.
- general code duplication. The most prominent example is value conversion.

Before the next review iteration I would like to see at least those cleaned up.

File: client/executor/wasmi/src/lib.rs [hunk 3]
```diff
@@ -155,30 +213,76 @@ impl<'a> Sandbox for FunctionExecutor<'a> {
...
-			val_ptr.into(),
-			&sandboxed_memory,
-			offset as usize,
-			val_len as usize,
-		) {
-			Ok(()) => Ok(sandbox_primitives::ERR_OK),
-			Err(_) => Ok(sandbox_primitives::ERR_OUT_OF_BOUNDS),
+			#[cfg(feature = "wasmer-sandbox")]
+			sandbox::Memory::Wasmer(sandboxed_memory) => {
+				let len = val_len as usize;
```
Comment (kpp) [inline client/executor/wasmi/src/lib.rs:238]: These `memory_set` and `memory_get` are very similar. Looks like you can reduce copypaste with a generic code. Also I would like to know why you could not use `MemoryInstance::transfer` to access the memory.

File: client/executor/wasmi/src/lib.rs [hunk 4]
```diff
@@ -261,13 +390,47 @@ impl<'a> Sandbox for FunctionExecutor<'a> {
...
+	type SupervisorFuncRef = wasmi::FuncRef;
+	type SC = FunctionExecutor;
+
+	fn with_sandbox_capabilities<R, F: FnOnce(&mut Self::SC) -> R>(f: F) -> R {
+		assert!(EXECUTOR.is_set(), "wasmi executor is not set");
+		EXECUTOR.with(|executor| f(&mut executor.clone()))
+	}
+}
+
+struct ThunkHolder;
```
Comment (kpp) [inline client/executor/wasmi/src/lib.rs:417]: Comments please

File: client/executor/wasmi/src/lib.rs [hunk 5]
```diff
@@ -261,13 +390,47 @@ impl<'a> Sandbox for FunctionExecutor<'a> {
...
+			.inner
+			.sandbox_store
+			.borrow()
 			.instance(instance_idx)
 			.map(|i| i.get_global_val(name))
 			.map_err(|e| e.to_string())
 	}
 }
 
+struct CapsHolder;
```
Comment (kpp) [inline client/executor/wasmi/src/lib.rs:403]: Comments please

File: client/executor/common/src/sandbox.rs [hunk 6]
```diff
@@ -332,34 +373,115 @@ impl<FR> SandboxInstance<FR> {
...
+							let wasmi_result = wasmi_instance
+								.invoke_export(export_name, args, guest_externals)?;
+							
+							Ok(wasmi_result)
+						}
+
+						#[cfg(feature = "wasmer-sandbox")]
+						BackendInstance::Wasmer(wasmer_instance) => {
+							let function = wasmer_instance
+								.exports
```
Comment (kpp) [inline client/executor/common/src/sandbox.rs:410]: I would suggest adding a trait `Backend` with functions `invoke_export`...

File: client/executor/common/src/sandbox.rs [hunk 7]
```diff
@@ -332,34 +373,115 @@ impl<FR> SandboxInstance<FR> {
...
+
+							if wasmer_result.len() > 1 {
+								return Err(wasmi::Error::Function(
+									"multiple return types are not supported yet".to_owned())
+								);
+							}
+
+							let wasmer_result = if let Some(wasmer_value) = wasmer_result.first() {
+								let wasmer_value = match *wasmer_value {
+									wasmer::Val::I32(val) => RuntimeValue::I32(val),
```
Comment (kpp) [inline client/executor/common/src/sandbox.rs:435]: sp_wasm_interface::Value and RuntimeValue are very alike. Is there a way to generalize type conversion?

File: client/executor/wasmtime/src/host.rs [hunk 8]
```diff
@@ -167,30 +168,72 @@ impl<'a> Sandbox for HostContext<'a> {
...
+				// This is safe because we construct slice from the same parts as
+				// the memory region itself. Current implementation is single threaded,
+				// so we should not face any synchronization or aliasing issues.
+				let source = unsafe { core::slice::from_raw_parts(
+					sandboxed_memory.data_ptr(),
+					sandboxed_memory.data_size() as usize)
+				};
+
+				self.inner.instance
+					.write_memory_from(
```
Comment (kpp) [inline client/executor/wasmtime/src/host.rs:228]: Here we go again, the same pattern to access memory which can be generalized.

**Review by kpp — CHANGES_REQUESTED** [review]

File: client/executor/common/src/sandbox.rs [hunk 9]
```diff
@@ -532,20 +752,9 @@ impl<FR> Store<FR> {
...
-				Pages(initial as usize),
-				maximum,
-			)?;
-
-		let mem_idx = self.memories.len();
-		self.memories.push(Some(mem));
-		Ok(mem_idx as u32)
+		let memories = &mut self.memories;
+		let backend_context = &self.backend_context;
+		Self::allocate_memory(memories, backend_context, initial, maximum).map(|(index, _)| index)
```
[the remark this answers is not in the document]
Comment (0x7CFE) [reply client/executor/common/src/sandbox.rs:757]: Good catch. If I'm not mistaken previously there were several places where this function was called.

File: client/executor/wasmi/src/lib.rs [hunk 10]
```diff
@@ -126,26 +133,52 @@ impl<'a> FunctionContext for FunctionExecutor<'a> {
...
+			#[cfg(feature = "wasmer-sandbox")]
+			sandbox::Memory::Wasmer(sandboxed_memory) => {
+				let len = buf_len as usize;
+				let mut buffer = vec![0; len];
+
+				if let Err(_) = sandboxed_memory.read_into(Pointer::new(offset as u32), &mut buffer) {
+					return Ok(sandbox_primitives::ERR_OUT_OF_BOUNDS)
+				}
+
+				todo!();
```
Comment (kpp) [inline client/executor/wasmi/src/lib.rs:173]: TODO

File: client/executor/wasmtime/src/host.rs [hunk 11]
```diff
@@ -200,42 +243,85 @@ impl<'a> Sandbox for HostContext<'a> {
...
+				};
+
+				let dst_range = match util::checked_range(offset as usize, len, sandboxed_memory.data_size() as usize) {
+					Some(range) => range,
+					None => return Ok(sandbox_primitives::ERR_OUT_OF_BOUNDS),
+				};
+
+				// This is safe because we construct slice from the same parts as
+				// the memory region itself. Current implementation is single threaded,
+				// so we should not face any synchronization or aliasing issues.
```
[the remark this answers is not in the document]
Comment (0x7CFE) [reply client/executor/wasmtime/src/host.rs]: done

**Comment by 0x7CFE** [comment]
OK, so I've addressed most of the review suggestions and redesigned the sandbox memory access API. It is now currently uses external buffers to avoid potential aliasing issues. Also a trait was introduced to abstract over backends.

File: client/executor/wasmtime/src/instance_wrapper.rs [hunk 12]
```diff
@@ -332,13 +345,13 @@ impl InstanceWrapper {
 	/// Write data to a slice of memory.
```
Comment (kpp) [inline client/executor/wasmtime/src/instance_wrapper.rs:345]:
```suggestion
	/// Write data to a pointer of memory from a slice of data.
```

**Comment by kpp** [comment]
LGTM, however I would like more experienced members to review this PR.

**Review by kpp — APPROVED** [review]

**Comment by 0x7CFE** [comment]
We also should refactor sandbox logic to eliminate copy-pasta in Value conversion. Yet, I'd really like to do that in a separate PR.

**Comment by athei** [comment]
This branch breaks the at least the `instr_local_get` benchmark of the contracts pallet:

```
Error: Input("Error executing runtime benchmark: Other(\"Wasm execution trapped: wasm trap: unreachable
wasm backtrace:
0: 0x361a - <unknown>!rust_begin_unwind
1: 0x2d31 - <unknown>!core::panicking::panic_fmt::h0dfe153eb0ef456a
2: 0x2a5d - <unknown>!core::result::unwrap_failed::ha926e221367dcf5b
3: 0x27c08c - <unknown>!<pallet_contracts::benchmarking::sandbox::Sandbox as core::convert::From<&pallet_contracts::benchmarking::code::WasmModule<T>>>::from::hba5f7fe780486c1d
4: 0x2705ca - <unknown>!<pallet_contracts::benchmarking::SelectedBenchmark as frame_benchmarking::utils::BenchmarkingSetup<T>>::instance::h3929c331f09dfe10
5: 0x3d1339 - <unknown>!pallet_contracts::benchmarking::<impl frame_benchmarking::utils::Benchmarking<frame_benchmarking::utils::BenchmarkResults> for pallet_contracts::pallet::Pallet<T>>::run_benchmark::{{closure}}::hfc055e856b03debc
6: 0x1124b1 - <unknown>!pallet_contracts::benchmarking::<impl frame_benchmarking::utils::Benchmarking<frame_benchmarking::utils::BenchmarkResults> for pallet_contracts::pallet::Pallet<T>>::run_benchmark::hd899e7824c309cb1
7: 0xfe64b - <unknown>!<node_runtime::Runtime as frame_benchmarking::utils::runtime_decl_for_Benchmark::Benchmark<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,u32>,node_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_spec_version::CheckSpecVersion<node_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<node_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<node_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<node_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<node_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<node_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<node_runtime::Runtime>)>>>>::dispatch_benchmark::h49bccb57517b7004
8: 0x2418ba - <unknown>!Benchmark_dispatch_benchmark\\nnote: run with `WASMTIME_BACKTRACE_DETAILS=1` environment variable to display more information\\n\")")
```

To reproduce run this command:
```
cargo run --release --features runtime-benchmarks,wasmer-sandbox -- benchmark --dev --execution=wasm --wasm-execution=compiled --heap-pages 4096 -p pallet_contracts -e instr_local_get
```

When you remove the `wasmer-sandbox` feature the bechmark does not trap.

This is the wasm module in question which triggers the trap. Note that this is not necessarily the exact module which generated the above error because during benchmark the amount of locals is scaled up but this is the module when the benchmark sets the amount of locals to the maximum:

[instr_local_get_1000.wasm.gz](https://github.com/paritytech/substrate/files/6800519/instr_local_get_1000.wasm.gz)

**Comment by 0x7CFE** [comment]
@athei, thank you very much for the input! I will have a look. As for the bench, is it possible to add extra `wasmer-sandbox` run to the CI?

**Comment by athei** [comment]
> As for the bench, is it possible to add extra `wasmer-sandbox` run to the CI?

Possible, yes. But do we want that. Then we are no longer testing the currently in use non-wasmer code path. We would need to do that on-top, right?

**Comment by 0x7CFE** [comment]
> > As for the bench, is it possible to add extra `wasmer-sandbox` run to the CI?
> 
> Possible, yes. But do we want that. Then we are no longer testing the currently in use non-wasmer code path. We would need to do that on-top, right?

I was talking about two code paths: one with and one without `wasmer-sandbox`.

**Comment by athei** [comment]
> > > As for the bench, is it possible to add extra `wasmer-sandbox` run to the CI?
> > 
> > 
> > Possible, yes. But do we want that. Then we are no longer testing the currently in use non-wasmer code path. We would need to do that on-top, right?
> 
> I was talking about two code paths: one with and one without `wasmer-sandbox`.

I don't think it is possible to just run the `pallet_contracts` (benchmark) tests with this feature enabled because there is no way to pass it through from `pallet_contracts`. One need to re-compile and test the whole workspace with this feature enabled.

I don't think the feature will influence the (benchmark) tests the way it is implemented right now, right? We would need a way to pass through the feature like this `pallet_contracts -> sp_sandbox -> ...`.

File: client/executor/common/src/sandbox.rs [hunk 13]
```diff
@@ -101,20 +123,28 @@ impl ImportResolver for Imports {
...
+			.as_wasmi()
+			.ok_or_else(|| {
+				wasmi::Error::Instantiation(format!(
+					"Unsupported non-wasmi export {}:{}",
+					module_name, field_name
+				))
+			})?;
+
+		// Here we use inner memory reference only to resolve
+		// the imports wthout accessing the memory contents.
```
Comment (athei) [inline client/executor/common/src/sandbox.rs:145]:
```suggestion
		// the imports without accessing the memory contents.
```

**Review by athei — APPROVED** [review]

**Comment by 0x7CFE** [comment]
> I would like to see in a follow up PR:
> 
>     * Automatically run `pallet_contract` tests with wasmer sandbox enabled (this includes the one generated from benchmarks).
> 
>     * Allow switching between wasmer and wasmi on a per instance basis. This could be useful to only allow audited contracts to run with wasmer at first.

Absolutely. As I wrote in the PR description:

> In some cases there might be an intention to force execution mode of a sandboxed code, say, for security or performance reasons. In that case command line parameter is needed. I tried adding one, but unfortunately this ended up being a non trivial task, since currently there is no way to pass configuration options to the sandbox environment which is instantiated on a per call basis. Of course, this is possible but would require substantial refactoring, which probably should be done in a separate PR.

**Comment by 0x7CFE** [comment]
> This branch breaks the at least the `instr_local_get` benchmark of the contracts pallet:
 
I've checked the errors and indeed, it gracefully fails because of too many local variables:

```
Validation error: locals exceed maximum (at offset 869)
```

Now trying to find out how to configure the limit. By the way, what limit is used in our benches and why? Was it based on a specification or was just set to "a big number"?

**Comment by 0x7CFE** [comment]
Ok, so this is a hard limit defined and checked in the wasmer source:

```rust
pub const MAX_WASM_FUNCTION_LOCALS: usize = 50000;

if self.num_locals > (MAX_WASM_FUNCTION_LOCALS as u32) {
    return Err(BinaryReaderError::new("locals exceed maximum", offset));
}
```

**Comment by athei** [comment]
> P.S.: Do we REALLY need that many locals? :)

It is a benchmark. It exercises the limits of what we can do in order to measure how long it takes. That said, it should **not** generate **that** many locals. It uses the stack height limit of the `Schedule` to determine how many locals to generate. I guess this benchmark is buggy. I will look into it.

**Comment by 0x7CFE** [comment]
The benchmark successfully passed after merging the fix :+1:

**Comment by 0x7CFE** [comment]
Postponed because of #9354 and #9431

**Review by TriplEight — APPROVED** [review]
LGTM.
If it can share any cache with other jobs, it would make sense to merge them.
Also, for the upcoming optimizations: can we run this test only if certain files were changed?

**Comment by athei** [comment]
> If it can share any cache with other jobs, it would make sense to merge them.

Adding the `wasmtime-sandbox` feature triggers a substantial amount of crates to be recompiled. I think having it as separate job and therefore being able to run it in parallel constitutes a lesser impact on CI waiting time.

> Also, for the upcoming optimizations: can we run this test only if certain files were changed?

In the future I hope that we can get rid of the `wasmer-sandbox` feature and can get this code to be always build. But if you implement that optimization we could add a job that actually executes this code which is to heavyweight to run for every commit.

**Comment by athei** [comment]
bot merge

**Comment by athei** [comment]
bot merge

File: client/executor/common/src/util.rs [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+				Ok(())
+			})
+		}
+
+		fn write_from(&self, dest_addr: Pointer<u8>, source: &[u8]) -> Result<()> {
+			self.0.with_direct_access_mut(|destination| {
+				let range = checked_range(dest_addr.into(), source.len(), destination.len())
+					.ok_or_else(|| Error::Other("memory write is out of bounds".into()))?;
+
+				&mut destination[range].copy_from_slice(source);
```
Comment (dvdplm) [inline client/executor/common/src/util.rs:108]: @0x7CFE This produces a warning on recent nightlies, `warning: unused borrow that must be used`. Not sure if you have any follow-up PRs coming; if so you can perhaps sneak in the `let _ = …` fix in one of them? :)

Comment (bkchr) [reply client/executor/common/src/util.rs:108]: This code doesn't make any sense at all. Why would you want to mutable borrow the result of `copy_from_slice`?

Comment (kpp) [reply client/executor/common/src/util.rs:108]: This borrow looks like a typo made during refactoring. The code is consistent with `read_into`, there is a guard check of the `range`, so the logic of this line is fine.

Comment (athei) [reply client/executor/common/src/util.rs:108]: Dmitry is on vacation. This indeed looks like an artifact. The rest of the function seems correct. Fix: #9615
